File: src/main/java/shopping/product/controller/ProductController.java
package shopping.product.controller;

import static shopping.util.ValidUtils.validNotEmpty;

import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import shopping.product.dto.request.ProductCreationRequest;
import shopping.product.dto.request.ProductUpdateRequest;
import shopping.product.dto.response.ProductResponse;
import shopping.product.service.ProductService;

@Controller
public class ProductController {

    private final ProductService productService;

    public ProductController(ProductService productService) {
        this.productService = productService;
    }

    @GetMapping("/")
    public String findAllProducts(Model model) {
        List<ProductResponse> productResponses = productService.readAllProduct();
        model.addAttribute("products", productResponses);
        return "index";
    }

    @GetMapping("/products/{productId}")
    @ResponseBody
    @ResponseStatus(HttpStatus.OK)
    public ProductResponse findProduct(@PathVariable Long productId) {
        return productService.readProduct(productId);
    }

    @PostMapping("/products")
    @ResponseBody
    @ResponseStatus(HttpStatus.OK)
    public ProductResponse createProduct(@RequestBody ProductCreationRequest productCreationRequest) {
        validCreationRequest(productCreationRequest);
        return productService.createProduct(productCreationRequest);
    }

    private void validCreationRequest(ProductCreationRequest productCreationRequest) {
        validNotEmpty(productCreationRequest.getImageUrl());
        validNotEmpty(productCreationRequest.getName());
        validNotEmpty(productCreationRequest.getPrice());
    }

    @PutMapping("/products/{productId}")
    @ResponseBody
    @ResponseStatus(HttpStatus.OK)
    public void updateProduct(
        @RequestBody ProductUpdateRequest productUpdateRequest,
        @PathVariable Long productId
    ) {
        validUpdateRequest(productUpdateRequest);
        productService.updateProduct(productUpdateRequest, productId);
    }

    private void validUpdateRequest(ProductUpdateRequest productUpdateRequest) {
        validNotEmpty(productUpdateRequest.getName());
        validNotEmpty(productUpdateRequest.getPrice());
        validNotEmpty(productUpdateRequest.getImageUrl());
    }

    @DeleteMapping("/products/{productId}")
    @ResponseBody
    @ResponseStatus(HttpStatus.OK)
    public void deleteProduct(@PathVariable Long productId) {
        productService.deleteProduct(productId);
    }
}

File: src/main/kotlin/shopping/product/controller/ProductController.kt
package shopping.product.controller

import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import shopping.common.PurgoMalumVerifier
import shopping.product.dto.response.ProductListResponseBody
import shopping.product.dto.request.ProductRequestBody
import shopping.product.dto.response.ProductResponseBody
import shopping.product.exception.ProductNotFoundException
import shopping.product.service.ProductService

@RestController
@RequestMapping("/api/products")
class ProductController(
    private val purgoMalumVerifier: PurgoMalumVerifier,
    private val productService: ProductService,
) {
    @GetMapping
    fun getProducts(): ProductListResponseBody =
        ProductListResponseBody(
            productService
                .findAll()
                .map(ProductResponseBody::from),
        )

    @GetMapping("/{id}")
    fun getProduct(
        @PathVariable id: Long,
    ): ProductResponseBody {
        val found = productService.findById(id) ?: throw ProductNotFoundException(id)
        return ProductResponseBody.from(found)
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun createProduct(
        @RequestBody @Valid requestBody: ProductRequestBody,
    ): ProductResponseBody {
        purgoMalumVerifier.containsProfanity(requestBody.name)
        val saved = productService.create(requestBody.toEntity())
        return ProductResponseBody.from(saved)
    }

    @PutMapping("/{id}")
    fun updateProduct(
        @PathVariable id: Long,
        @RequestBody @Valid requestBody: ProductRequestBody,
    ): ProductResponseBody {
        purgoMalumVerifier.containsProfanity(requestBody.name)
        val saved = productService.update(id, requestBody.toEntity())
        return ProductResponseBody.from(saved)
    }

    @DeleteMapping("/{id}")
    fun deleteProduct(
        @PathVariable id: Long,
    ): String {
        productService.deleteById(id)
        return "Product with id $id deleted successfully."
    }
}
